File: app/src/main/java/com/gmail/mostafa/ma/saleh/yawa/fragments/DetailsFragment.java
package com.gmail.mostafa.ma.saleh.yawa.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.lzyzsd.circleprogress.ArcProgress;
import com.gmail.mostafa.ma.saleh.yawa.R;
import com.gmail.mostafa.ma.saleh.yawa.models.Day;
import com.gmail.mostafa.ma.saleh.yawa.utilities.Constants;
import com.gmail.mostafa.ma.saleh.yawa.utilities.SharedPreferencesManager;
import com.gmail.mostafa.ma.saleh.yawa.utilities.StringUtils;
import com.gmail.mostafa.ma.saleh.yawa.utilities.Utils;

import java.util.Date;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DetailsFragment extends Fragment {

    private static final String PARAM_DAY = "param_day";
    private static final String PARAM_DATE = "param_date";

    @BindView(R.id.tv_temp_max)
    TextView tvTempMax;
    @BindView(R.id.tv_temp_min)
    TextView tvTempMin;
    @BindView(R.id.tv_weather_status)
    TextView tvStatus;
    @BindView(R.id.tv_humidity)
    TextView tvHumidity;
    @BindView(R.id.tv_wind_speed)
    TextView tvWindSpeed;
    @BindView(R.id.tv_wind_direction)
    TextView tvWindDirection;
    @BindView(R.id.tv_pressure)
    TextView tvPressure;
    @BindView(R.id.img_weather)
    ImageView imgWeatherStatus;
    @BindView(R.id.aprg_clouds)
    ArcProgress aprgClouds;

    private Day mDay;
    private Date mDate;

    public static DetailsFragment newInstance(Day day, Date date) {
        DetailsFragment fragment = new DetailsFragment();
        Bundle args = new Bundle();
        args.putSerializable(PARAM_DAY, day);
        args.putSerializable(PARAM_DATE, date);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mDay = (Day) getArguments().getSerializable(PARAM_DAY);
            mDate = (Date) getArguments().getSerializable(PARAM_DATE);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_details, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        ButterKnife.bind(this, view);
        if (getActivity() != null) {
            getActivity().setTitle(StringUtils.formatDate(Constants.LONG_DATE_PATTERN, mDate));
            tvTempMax.setText(StringUtils.format("%d°", Math.round(mDay.temp.max)));
            tvTempMin.setText(StringUtils.format("%d°", Math.round(mDay.temp.min)));
            tvStatus.setText(mDay.weather[0].main);
            tvHumidity.setText(mDay.humidity == 0.0 ? "N/A" : StringUtils.format("%d%%", Math.round(mDay.humidity)));
            imgWeatherStatus.setImageResource(Utils.getArtResourceForWeatherCondition(mDay.weather[0].id));
            tvPressure.setText(StringUtils.format("%d %s", Math.round(mDay.pressure), getString(R.string.pressure_unit)));
            tvWindSpeed.setText(StringUtils.format("%d %s", Math.round(mDay.speed), getWindSpeedUnit()));
            tvWindDirection.setText(getDirection(mDay.deg));
            aprgClouds.setProgress(Math.round(mDay.clouds));
        }
    }

    private String getDirection(float angle) {
        int val = (int) ((angle / 22.5) + 0.5);
        return getResources().getStringArray(R.array.directions)[val % 16];
    }

    private String getWindSpeedUnit() {
        SharedPreferencesManager preferencesManager = SharedPreferencesManager.getInstance();
        String tempUnits = preferencesManager.getString(Constants.KEY_TEMP_UNITS, "metric");
        if (tempUnits.contains("imperial")) {
            return getString(R.string.wind_speed_unit_imperial);
        } else {
            return getString(R.string.wind_speed_unit_metric);
        }
    }
}

File: app/src/main/java/com/gmail/mostafa/ma/saleh/yawa/fragments/DetailsFragment.kt
package com.gmail.mostafa.ma.saleh.yawa.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.gmail.mostafa.ma.saleh.yawa.R
import com.gmail.mostafa.ma.saleh.yawa.models.Day
import com.gmail.mostafa.ma.saleh.yawa.utilities.Constants
import com.gmail.mostafa.ma.saleh.yawa.utilities.PreferencesUtils
import com.gmail.mostafa.ma.saleh.yawa.utilities.ResourcesUtils
import com.gmail.mostafa.ma.saleh.yawa.utilities.StringUtils
import kotlinx.android.synthetic.main.fragment_details.*
import java.util.*

class DetailsFragment : Fragment() {

    private var mDay: Day? = null
    private var mDate: Date? = null

    private val windSpeedUnit: String
        get() {
            val tempUnits = PreferencesUtils.getString(Constants.KEY_TEMP_UNITS, "metric")
            return if (tempUnits.contains("imperial")) {
                getString(R.string.wind_speed_unit_imperial)
            } else {
                getString(R.string.wind_speed_unit_metric)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mDay = arguments?.getParcelable(PARAM_DAY)
        mDate = arguments?.getSerializable(PARAM_DATE) as? Date
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_details, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        activity?.title = StringUtils.formatDate(Constants.LONG_DATE_PATTERN, mDate!!)
        mDay?.run {
            temp_max_text_view.text = String.format("%d°", Math.round(temperature.max))
            temp_min_text_view.text = String.format("%d°", Math.round(temperature.min))
            weather_text_view.text = weather[0].main
            tv_humidity.text = if (humidity == 0) "N/A" else String.format("%d%%", humidity)
            weather_image_view.setImageResource(ResourcesUtils.getArtResourceForWeatherCondition(weather[0].id))
            tv_pressure.text = String.format("%d %s", Math.round(pressure), getString(R.string.pressure_unit))
            tv_wind_speed.text = String.format("%d %s", Math.round(windSpeed), windSpeedUnit)
            tv_wind_direction.text = getDirection(windDirection)
            clouds_arc_progress.progress = Math.round(clouds).toInt()
        }
    }

    private fun getDirection(angle: Double): String {
        val value = (angle / 22.5 + 0.5).toInt()
        return resources.getStringArray(R.array.directions)[value % 16]
    }

    companion object {

        private const val PARAM_DAY = "param_day"
        private const val PARAM_DATE = "param_date"

        fun newInstance(day: Day, date: Date) = DetailsFragment().apply {
            arguments = with(Bundle()) {
                putParcelable(PARAM_DAY, day)
                putSerializable(PARAM_DATE, date)
                return@with this
            }
            return this
        }
    }
}
